File: app/src/main/java/com/parabola/newtone/mvp/presenter/ChoosePlaylistPresenter.java
package com.parabola.newtone.mvp.presenter;

import com.parabola.domain.executor.SchedulerProvider;
import com.parabola.domain.repository.PlaylistRepository;
import com.parabola.newtone.di.app.AppComponent;
import com.parabola.newtone.mvp.view.ChoosePlaylistView;
import com.parabola.newtone.ui.router.MainRouter;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.internal.observers.CallbackCompletableObserver;
import moxy.InjectViewState;
import moxy.MvpPresenter;

@InjectViewState
public final class ChoosePlaylistPresenter extends MvpPresenter<ChoosePlaylistView> {

    @Inject PlaylistRepository playlistRepo;
    @Inject SchedulerProvider schedulers;

    @Inject MainRouter router;

    private final int[] trackIds;
    private final CompositeDisposable disposables = new CompositeDisposable();

    public ChoosePlaylistPresenter(AppComponent appComponent, int[] trackIds) {
        appComponent.inject(this);
        this.trackIds = trackIds;
    }

    @Override
    protected void onFirstViewAttach() {
        disposables.addAll(observePlaylistsUpdates());
    }

    @Override
    public void onDestroy() {
        disposables.dispose();
    }


    private Disposable observePlaylistsUpdates() {
        return playlistRepo.observePlaylistsUpdates()
                .flatMapSingle(o -> playlistRepo.getAll())
                .subscribeOn(schedulers.io())
                .observeOn(schedulers.ui())
                .subscribe(getViewState()::refreshPlaylists);
    }


    public void onClickCreateNewPlaylist() {
        router.openCreatePlaylistDialog();
    }

    public void onClickPlaylistItem(int playlistId) {
        playlistRepo.addTracksToPlaylist(playlistId, trackIds)
                .subscribeOn(schedulers.io())
                .observeOn(schedulers.ui())
                .subscribe(new CallbackCompletableObserver(getViewState()::closeScreen));
    }
}

File: app/src/main/java/com/parabola/newtone/presentation/chooseplaylist/ChoosePlaylistPresenter.kt
package com.parabola.newtone.presentation.chooseplaylist

import com.parabola.domain.executor.SchedulerProvider
import com.parabola.domain.repository.PlaylistRepository
import com.parabola.newtone.di.app.AppComponent
import com.parabola.newtone.ui.router.MainRouter
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.internal.observers.CallbackCompletableObserver
import moxy.InjectViewState
import moxy.MvpPresenter
import javax.inject.Inject

@InjectViewState
class ChoosePlaylistPresenter(
    appComponent: AppComponent,
    private val trackIds: IntArray,
) : MvpPresenter<ChoosePlaylistView>() {

    @Inject
    lateinit var playlistRepo: PlaylistRepository

    @Inject
    lateinit var schedulers: SchedulerProvider

    @Inject
    lateinit var router: MainRouter

    private val disposables = CompositeDisposable()


    init {
        appComponent.inject(this)
    }


    override fun onFirstViewAttach() {
        disposables.addAll(observePlaylistsUpdates())
    }

    override fun onDestroy() {
        disposables.dispose()
    }


    private fun observePlaylistsUpdates(): Disposable {
        return playlistRepo.observePlaylistsUpdates()
            .flatMapSingle { playlistRepo.all }
            .subscribeOn(schedulers.io())
            .observeOn(schedulers.ui())
            .subscribe(viewState::refreshPlaylists)
    }


    fun onClickCreateNewPlaylist() {
        router.openCreatePlaylistDialog()
    }

    fun onClickPlaylistItem(playlistId: Int) {
        playlistRepo.addTracksToPlaylist(playlistId, *trackIds)
            .subscribeOn(schedulers.io())
            .observeOn(schedulers.ui())
            .subscribe(CallbackCompletableObserver { viewState.closeScreen() })
    }

}
